Data Visualization with Chartify (#5957)
section-engineering-education/engineering-education: content/articles/data-viz-chartify/index.md

- Basic knowledge of Python.
- Basic knowledge in using Jupyter Notebooks or any other notebook-based technology. I would recommend Jupyter notebook or Google Colab for data visualization. However, for this tutorial, we will be using [Jupyter](https://jupyter.org/).

### What is Chartify?
According to the documentation, [Chartify](https://chartify.readthedocs.io/en/latest/) is a Python library that aims to make it easy for data scientists to create charts. This open-source library was introduced to the world in this article by Spotify Lab, [Introducing Chartify: Easier chart creation in Python for data scientists](https://engineering.atspotify.com/2018/11/15/introducing-chartify-easier-chart-creation-in-python-for-data-scientists/), as a means for you and me to build visually appealing charts using Python.

The library was built on Bokeh. Bokeh is a Python visualization framework that JavaScript based. Bokeh can be used to generate interactive visualizations for modern web browsers. You can use Bokeh to generate JavaScript-powered visualizations without having to write any JavaScript. I am sure you know what that means for Chartify... Yeah, you got it! Fantastic visuals.

> ![Introduction to Chartify](/engineering-education/data-viz-chartify/chartifydoc.png)

Reviewer: ```suggestion
![Introduction to Chartify](/engineering-education/data-viz-chartify/chartifydoc.png)

```

If you are working on your local machine, you will need to install Chartify. I highly recommend creating a virtual environment to better manage your packages. 

Chartify can be installed by running the code below:

```python 
pip install chartify
```
If you're using the Jupyter notebook and want to install a package with pip, you might have tied to run pip directly in the shell. However, this might not generally work. Thus, you can use the code snippet below:

```python
Import sys
!{sys.executale} -m pip install chartify
```
Chartify requires a ChromeDriver. The library requires this to generate a PNG output. You can download ChromeDriver from [the ChromeDriver for Chrome Website](https://sites.google.com/a/chromium.org/chromedriver/home). After downloading, you need to install it and copy it into the appropriate directory. This article will show you [how to add executables to your Windows PATH](https://medium.com/@kevinmarkvi/how-to-add-executables-to-your-path-in-windows-5ffa4ce61a53).

### Importing Chartify

```python 
import numpy as np
import pandas as pd
import chartify
```

> Pandas is a python library that is required for data cleaning and manipulation while Numpy is a python library for scientific computing. 

Reviewer: Change python to Python for  all instances used in this tutorial

### Loading the dataset

Chartify comes with an in-built dataset. Run the code below to access it. 

```python
data = chartify.examples.example_data()
data.head()
```

> However, for this tutorial, we are going to make use of the [complete pokemon dataset](https://www.kaggle.com/rounakbanik/pokemon) which you can download for free from Kaggle. After unzipping the file, you should create a python file on your local machine in the same folder as your dataset. 

Reviewer: ```suggestion
However, for this tutorial, we are going to make use of the [complete Pokemon dataset](https://www.kaggle.com/rounakbanik/pokemon) which you can download for free from Kaggle. After unzipping the file, you should create a python file on your local machine in the same folder as your dataset. 
```

Reviewer: Change pokeman to Pokeman
Do this for all instances in this article

This dataset was also used in this tutorial piece on [Seaborn](https://www.section.io/engineering-education/seaborn-tutorial/). We would reproduce some of the graphs from the article and add other charts.

Now let us read and display the dataset by running the code below.

```python
pokemon_df = pd.read_csv('pokemon.csv')
pokemon_df.head(5)
```
We will check our column using the `.columns` and add an `s/n` column using the code snippet below.

```python
pokemon_df.insert(loc =0, column = 's/n',value = np.arange(len(pokemon_df)))
```
### Problem Encounter while using Chartify
While working with Charify I discovered a problem. I was unable to display my chart. 

> when I ran my codes for the first time, I got an error message. If we take a look at the error message, we will see it was a compatibility issue. The latest version of ChromeDriver available only supports Chrome version 93 and my current Chrome browser version is 96.0.4664.45. 

Reviewer: ```suggestion
When I ran my codes for the first time, I got an error message. If we take a look at the error message, we will see it was a compatibility issue. The latest version of ChromeDriver available only supports Chrome version 93 and my current Chrome browser version is 96.0.4664.45. 
```

Reviewer: Start with a capital letter

- **Step 2**: Download Older Chrome Version. You can get this from sites like [FileHippo](https://www.slimjet.com/chrome/google-chrome-old-version.php) and [SlimJet](https://filehippo.com/download_google-chrome/history/) since there is an official repository hosting an older build of the Chrome browser.
- **Step 3**:  Disable Chrome Auto-Updates. 

You can get a more detailed step-by-step tutorial in this article, [How to Downgrade and Install Older Versions of Chrome?](https://browserhow.com/how-to-downgrade-and-install-older-version-of-chrome/).

### Data Visualization using Chartify

To get started with our charts plotting, type and run the below code snippet.

```python
ch = chartify.Chart()
ch.show()
```
The code we just ran will give us the output below. 

![Chartify](/engineering-education/data-viz-chartify/chartifyone.png)
*Screenshot by author*

Chartify makes it relatively easy for first-timers to get started. That's one amazing thing about it. This empty chart that got displayed shows users of the library how they can fill the chart with data. 

Let�s visualize the Pokemon type distribution using a bar chart.

### Pokemon type distribution

> A bar chart represents data by using bars of different heights. We would use on to see the distribution of the various pokemon types. First, we need to create a data frame by grouping the Pokemon type 1 and its generation. This can be done using the code snippet below:

Reviewer: ```suggestion
A bar chart represents data by using bars of different heights. We would use on to see the distribution of the various Pokemon types. First, we need to create a data frame by grouping the Pokemon type 1 and its generation. This can be done using the code snippet below:
```

```python
bar_data = (pokemon_df.groupby('type1')[['generation']].sum()
            .reset_index()
           )
bar_data
```
Next, we would visualize this data frame. 

```python
ch = chartify.Chart(blank_labels=True, x_axis_type='categorical')
ch.set_title("Pokemon EDA using Chartify")
ch.set_subtitle("Pokemon type distribution")
ch.plot.bar(
        data_frame=bar_data,
        categorical_columns=['type1'],
        numeric_column='generation')
ch.show()
```
The first few arguments that we passed are `.set_title` and `.set_subtitle` which are for your chart title and subtitle. You can specify a name for your x and y axes using `.axes.set_xaxis_label` and `.axes.set_yaxis_label` respectively. The last argument, `.plot.bar` requires your data_frame which is the data to be plotted. The other parameters define the categorical and numerical columns you want to plot. 

> ![Chartify](/engineering-education/data-viz-chartify/chartifytwo.png)

Reviewer: ```suggestion
![Chartify](/engineering-education/data-viz-chartify/chartifytwo.png)

```

*Screenshot by author*

Next, let�s look at the pokemon type distribution but this time taking account of the �is_legendary� attribute of each pokemon. We would also start by creating a dataframe of the data needed. We can achieve this by grouping our pokemon by the �type1� and �is_legendary� columns. This can be done by running the code snippet below:

```python
Pokemon_type_distribution = (pokemon_df.groupby(['type1','is_legendary'])['s/n'].sum().reset_index())
print(Pokemon_type_distribution.head())
```

> Then we would visualize this data with Chartify by running:

Reviewer: ```suggestion
Then we would visualize this data with Chartify by running code below:
```

ch.style.color_palette.reset_palette_order()
ch.plot.text(
        data_frame=generations,
        categorical_columns='generation',
        numeric_column='count',
        text_column='count',
        color_column='generation')
ch.show()
```
We did not discuss the `color_column` in the previous bar chart. The `color_column` parameter gives your plot a color-based argument that colors the plot of the basis of the column specified. In our graph, Chartify will assign different colors to the different generations because we set `color_column` to be the column �generation�.

![Chartify](/engineering-education/data-viz-chartify/chartifyfour.png)
*Screenshot by author*

### Number of Pokemon of each type in each generation

Let us try out a line graph. We would use our knowledge of Python to extract our data from our pokemon data. We would start by getting a count of each type.

```python
data2 = pokemon_df[pokemon_df.type2.notna()].copy()
data2.type1 = data2.type2
type_data = pd.concat ([pokemon_df, data2])

#get a count for each type
type_count = type_data.type1.value_counts()
type_count = pd.DataFrame ({'count': type_count})
type_count
```

> Next, we would create a dataset that has the total for each pokemon type for each generation by using the `.groupby` and `.value_counts` function.

Reviewer: ```suggestion
Next, we wil create a dataset that has the total for each Pokemon type for each generation by using the `.groupby` and `.value_counts` function.
```

```python
generations_types = type_data.groupby (['generation']).type1.value_counts()
generations_types = pd.DataFrame ({'count': generations_types})
generations_types['generation'] = generations_types.index.get_level_values(0)
generations_types['type'] = generations_types.index.get_level_values(1)
generations_types.reset_index(drop = True, inplace = True)
generations_types
```
Now we visualize our data by plotting the line graph using chartify `.plot.parallel` function.

```python
ch = chartify.Chart(blank_labels=True, x_axis_type='categorical')
ch.set_title("Pokemon EDA using Chartify")
ch.set_subtitle("Number of Pokemon of Each Type in Each Generation")
ch.plot.parallel(
        data_frame=generations_types,
        categorical_columns='generation',
        numeric_column='count',
        color_column='type')
ch.show()
```
Just like the bar charts, the line graph takes the in a data_frame argument for our data. We will also specify the `categorical_column`, `numeric_column`, and `color_column` in the same way as it did in the bar chart. After that, we run the code, and our output is displayed.

> ![Chartify](/engineering-education/data-viz-chartify/chartifyfive.png)

Reviewer: ```suggestion
![Chartify](/engineering-education/data-viz-chartify/chartifyfive.png)

```

Reviewer: Put an empty line below image
Do this for all images in this article

        numeric_column='count',
        color_column='generation')
ch.axes.set_xaxis_tick_orientation('horizontal')
ch.show()
```

![Chartify](/engineering-education/data-viz-chartify/chartifysix.png)
*Screenshot by author*

### Pokemon Type Combination

We would plot a heatmap to see the pokemon type combination. Just as we did for the other visualizations, we would create a data frame. 

```Python
type_combos= (pokemon_df.groupby(['type1', 'type2'])['generation'].mean().reset_index())
```
And we would use the `.plot.heatmap` to create the heatmap. 

```python
ch = chartify.Chart(blank_labels=True, x_axis_type='categorical', y_axis_type='categorical')
ch.plot.heatmap(
        data_frame=type_combos,
        x_column='type1',
        y_column='type2',
        color_column='generation',
        text_column='generation',
        text_color='white')
ch.axes.set_xaxis_label('type1')
ch.axes.set_yaxis_label('type2')
ch.set_title('Pokemon EDA using Chartify')
ch.set_subtitle("Pokemon Type Combinations: Heatmap")
ch.show()
```

![Chartify](/engineering-education/data-viz-chartify/chartifyseven.png)
*Screenshot by author*

### Pokemon Weight distribution

> Let�s have a look at our pokemon weight. This will best be displayed in a histogram. In the �weight_kg� column, there were nan values. We will have to fill it. For this article, we would fill it with 0.

Reviewer: ```suggestion
Let�s have a look at our pokemon weight. This will best be displayed in a histogram. In the `weight_kg` column, there were nan values. We will have to fill it. For this article, we would fill it with `0`.
```

Reviewer: Also use back ticks here

```python
pokemon_df['weight_kg'] = pokemon_df['weight_kg'].fillna(0)
```

After that, we will plot our histogram.

```python
ch = chartify.Chart(blank_labels=True, y_axis_type='density')
ch.set_title("Pokemon EDA using Chartify")
ch.set_subtitle("Distribution of Weight: Histogram")
ch.plot.histogram(data_frame=pokemon_df, values_column='weight_kg',bins=50)
ch.show()
```
![Chartify](/engineering-education/data-viz-chartify/chartifyeight.png)
*Screenshot by author*

### Conclusion

To gain insight into their vast amounts of data, data visualization is vital. And Chartify ensures you get it done with less code. We hope that this post will help you get a feel of how you can use Chartify for data visualization. 

Finally, if you want to take a look at the code used in this post, you can find it [here](https://github.com/Bennykillua/Project/blob/main/Data%20visualization%20using%20Charify/Chartify.ipynb).

Happy coding...

### Resources
- [Welcome to chartify�s documentation!](https://chartify.readthedocs.io/en/latest/)

> - [Chartify data visualization](https://www.geeksforgeeks.org/data-visualisation-with-chartify/)

Reviewer: ```suggestion
- [Chartify data visualization](https://www.geeksforgeeks.org/data-visualisation-with-chartify/)

---
Peer Review Contributions by: [Bonface Muriithi](/engineering-education/authors/bonface-muriithi)
```

Reviewer: Added PRL

```python
bar_data = (pokemon_df.groupby('type1')[['generation']].sum()
            .reset_index()
           )
bar_data
```
Next, we would visualize this data frame. 

```python
ch = chartify.Chart(blank_labels=True, x_axis_type='categorical')
ch.set_title("Pokemon EDA using Chartify")
ch.set_subtitle("Pokemon type distribution")
ch.plot.bar(
        data_frame=bar_data,
        categorical_columns=['type1'],
        numeric_column='generation')
ch.show()
```
The first few arguments that we passed are `.set_title` and `.set_subtitle` which are for your chart title and subtitle. You can specify a name for your x and y axes using `.axes.set_xaxis_label` and `.axes.set_yaxis_label` respectively. The last argument, `.plot.bar` requires your data_frame which is the data to be plotted. The other parameters define the categorical and numerical columns you want to plot. 

![Chartify](/engineering-education/data-viz-chartify/chartifytwo.png)

*Screenshot by author*

> Next, let’s look at the pokemon type distribution but this time taking account of the ‘is_legendary’ attribute of each Pokemon. We would also start by creating a dataframe of the data needed. We can achieve this by grouping our Pokemon by the “type1” and “is_legendary” columns. This can be done by running the code snippet below:

Reviewer: ```suggestion
Next, let’s look at the Pokemon type distribution but this time taking account of the `is_legendary` attribute of each Pokemon. We would also start by creating a data-frame of the data needed. We can achieve this by grouping our Pokemon by the `type1` and `is_legendary` columns. This can be done by running the code snippet below:
```

Reviewer: Use back ticks when using names included in the codes

        numeric_column='s/n', color_column='is_legendary')
ch.show()
```
![Chartify](/engineering-education/data-viz-chartify/chartifythree.png)

*Screenshot by author*

### Number of Pokemon in each generation

Let's take a look at the number of Pokemon in each generation. We would extract this data from the Pokemon data by running the code snippet below.

```python
generations = pd.DataFrame ({'count': pokemon_df.generation.value_counts().sort_index()})
generations['generation'] = generations.index.get_level_values(0)
generations
```

Next, we would run a code snippet like the one we ran for the previous bar chart.

```python
ch = chartify.Chart(blank_labels=True, x_axis_type='categorical')
ch.set_title("Pokemon EDA using Chartify")
ch.set_subtitle("Number of Pokemon in Each Generation")
ch.plot.bar(data_frame=generations, categorical_columns='generation', 
                             numeric_column='count',color_column='generation')
ch.style.color_palette.reset_palette_order()
ch.plot.text(
        data_frame=generations,
        categorical_columns='generation',
        numeric_column='count',
        text_column='count',
        color_column='generation')
ch.show()
```

> We did not discuss the `color_column` in the previous bar chart. The `color_column` parameter gives your plot a color-based argument that colors the plot of the basis of the column specified. In our graph, Chartify will assign different colors to the different generations because we set `color_column` to be the column “generation”.

Reviewer: ```suggestion
The `color_column` parameter gives your plot a color-based argument that colors the plot of the basis of the column specified. In our graph, Chartify will assign different colors to the different generations because we set `color_column` to be the column `generation`.
```

Reviewer: Remove this statement

![Chartify](/engineering-education/data-viz-chartify/chartifyfour.png)

*Screenshot by author*

### Number of Pokemon of each type in each generation

> Let us try out a line graph. We would use our knowledge of Python to extract our data from our pokemon data. We would start by getting a count of each type.

Reviewer: ```suggestion
Let us try out a line graph. We would use our knowledge of Python to extract our data from our Pokemon data. We would start by getting a count of each type.
```

---
layout: engineering-education
status: publish
published: true
url: /data-viz-chartify/
title: Data Visualization with Chartify
description: Chartify is an open-source data visualization library from Spotify that makes it easy for data analysts to create charts and graphs. This article gives a brief introduction to this technology. 
author: benedict-ifeanyi-iheagwara
date: 2022-01-18T00:00:00-00:20
topics: [API]
excerpt_separator: <!--more-->
images:
  - url: /engineering-education/data-viz-chartify/hero.png
    alt: Data visualisation with Chartify
---

> Although Chartify isn�t the new kid on the block, many data scientists are still unfamiliar with it. That�s a bummer, considering that it was developed to make data visualization easier for data scientists and is more 'intuitive' than other Python charting libraries. Data visualization is an important aspect of data analytics since it allows you to communicate information to shareholders in a visually appealing and understandable way.

Reviewer: ```suggestion
Although Chartify isn't the new kid on the block, many data scientists are still unfamiliar with it. That's a bummer, considering that it was developed to make data visualization easier for data scientists and is more 'intuitive' than other Python charting libraries. Data visualization is an important aspect of data analytics since it allows you to communicate information to shareholders in a visually appealing and understandable way.
```

Reviewer: Ops I forgot  to highlight this minor change @Bennykillua
